### DISCUSSION TITLE
harbor portal with path based routing (/harbor) instead of root (/)

### DISCUSSION THREAD

**Discussion by sky29** [goharbor/harbor-helm#1323] [Q&A]
Hello,

I am trying to install harbor using "harbor-helm" chart. I want to serve "harbor portal" on custom path e.g. ( https://monitoring.example.com/harbor/ ), but it is not working & throwing errors.

My stack looks like this: 
1. GKE cluster + custom nginx proxy (with custom nginx config and daemonset, not ingress) + https (which will be terminated at nginx level) + "harbor-helm" chart.

2. Harbor Helm Chart values.yaml looks like this:
```
expose:
    type: ClusterIP
  tls:
    enabled: false
externalURL: "http://monitoring.example.com/harbor/"
```

After helm install, when I port-forwarded the service and harbor-portal is loading in browser properly, means service is running fine at root url ("/").
kubectl port-forward svc/harbor-portal -n harbor 7000:80

3. My Nginx server and location context (with path based routing) looks like this:

```
server {
    listen 80 default_server;
    server_name _;
    return 301 https://$host$request_uri;
}
server {
    listen 443 default_server ssl;
    server_name  monitoring.example.com;
    resolver kube-dns.kube-system.svc.cluster.local valid=10s;
    ssl_certificate     /etc/tls.crt;
    ssl_certificate_key /etc/tls.key;
    
    location /harbor {
        rewrite ^/harbor/(.*)$ /$1 break;
        proxy_pass      http://harbor-portal.harbor.svc.cluster.local;
        proxy_set_header Host  $host;
        proxy_set_header X-Forwarded-For $remote_addr;
   }
   
}
```

4. But when I navigate to url (https://monitoring.example.com/harbor/) in browser, I am getting blank white page with text "Loading..." & error in browser's developer console something like this:

light-theme.css?buildTimestamp=1661659968787 net::ERR_ABORTED 404 (Not Found)

looks like harbor-portal angular app base path url rewriting not working properly with nginx.

5. same nginx "server and location context" is working fine for other solutions (like grafana, prometheus etc.)

6. I saw similar post:
https://github.com/goharbor/harbor/issues/16031

I tried nginx configuration mentioned there (ingress doesn't apply to me), but it is not working for me.

Anyone can help, how to serve "harbor-portal" through custom path based routing (like "/harbor" instead of root "/") ?

**Comment by amrap030** [comment]
I have the exact same problem, have you found a solution?

**Comment by alexchiri** [comment]
This is what worked for me:

```
    worker_processes auto;
    pid /tmp/nginx.pid;
    events {
        worker_connections  1024;
    }
    http {
        client_body_temp_path /tmp/client_body_temp;
        proxy_temp_path /tmp/proxy_temp;
        fastcgi_temp_path /tmp/fastcgi_temp;
        uwsgi_temp_path /tmp/uwsgi_temp;
        scgi_temp_path /tmp/scgi_temp;
        server {
            listen 8080;
            listen [::]:8080;
            server_name  localhost;
            root   /usr/share/nginx/html;
            index  index.html index.htm;
            include /etc/nginx/mime.types;
            gzip on;
            gzip_min_length 1000;
            gzip_proxied expired no-cache no-store private auth;
            gzip_types text/plain text/css application/json application/javascript application/x-javascript text/xml application/xml application/xml+rss text/javascript;
            location /devcenter-api-2.0 {
                try_files $uri $uri/ /swagger-ui-index.html;
            }
            location /harbor {
                rewrite ^/harbor/(.*) /$1 break;
                try_files $uri $uri/ /index.html;
            }
            location = /index.html {
                add_header Cache-Control "no-store, no-cache, must-revalidate";
                sub_filter 'base href="/"' 'base href="/harbor/"';
                sub_filter_once on;
            }
        }
    }
```

And also not forgetting to delete the harbor-portal pod in order for the changes in the configmap to become effective.

**Reply by alexchiri** [reply]
But then I got stuck in login, because the login form tries to go the endpoint without /harbor/ in path. I think harbor is simply not made to be run on a path.

**Reply by hAislt** [reply]
> This is what worked for me:
> 
> ```
>     worker_processes auto;
>     pid /tmp/nginx.pid;
>     events {
>         worker_connections  1024;
>     }
>     http {
>         client_body_temp_path /tmp/client_body_temp;
>         proxy_temp_path /tmp/proxy_temp;
>         fastcgi_temp_path /tmp/fastcgi_temp;
>         uwsgi_temp_path /tmp/uwsgi_temp;
>         scgi_temp_path /tmp/scgi_temp;
>         server {
>             listen 8080;
>             listen [::]:8080;
>             server_name  localhost;
>             root   /usr/share/nginx/html;
>             index  index.html index.htm;
>             include /etc/nginx/mime.types;
>             gzip on;
>             gzip_min_length 1000;
>             gzip_proxied expired no-cache no-store private auth;
>             gzip_types text/plain text/css application/json application/javascript application/x-javascript text/xml application/xml application/xml+rss text/javascript;
>             location /devcenter-api-2.0 {
>                 try_files $uri $uri/ /swagger-ui-index.html;
>             }
>             location /harbor {
>                 rewrite ^/harbor/(.*) /$1 break;
>                 try_files $uri $uri/ /index.html;
>             }
>             location = /index.html {
>                 add_header Cache-Control "no-store, no-cache, must-revalidate";
>                 sub_filter 'base href="/"' 'base href="/harbor/"';
>                 sub_filter_once on;
>             }
>         }
>     }
> ```
> 
> And also not forgetting to delete the harbor-portal pod in order for the changes in the configmap to become effective.

I have the exact same issue.
It's just sad that close to all helm charts can do this as there is something like baseurl and the routes are according to that configuration.

After each deployment the configmap will get overwritten anyway how do you make this change persistent?

edit: I fiddled with ingressroute as the API endpath is not working. I don't think we will be able to solve this without recompiling the docker image.
